### PR TITLE
[BSE-4811] Validate left_on and right_on in Merge

### PR DESCRIPTION
**Pull request by chrisoh7**
## Changes included in this PR

Adds Validation methods for on, left_on and right_on parameters in merge. 

Checks for type correctness: currently only str, str list, str tuple, or None are supported. 
Number of elements in left_on and right_on should match. If failed to validate, raises error.
Checks membership in left and right DFs to validate keys.
Also disallows accepting both on and left_on && right_on, should have one of the two (or none)

For future consideration:
* Addressing bug when no arguments are put in (merge should work-automatically collect common cols, but currently raising errors)
* Also might have to double check and update validation function once on-argument support is added 

## Testing strategy

PR CI, adding merge tests with left_on and right_on of varying length/types, None values, etc.

## User facing changes

Support checking for Invalid inputs for left_on and right_on in merge. 

## Checklist
- [x] Pipelines passed before requesting review. To run CI you must include `[run CI]` in your commit message.
- [x] I am familiar with the [Contributing Guide](https://github.com/bodo-ai/Bodo/blob/main/CONTRIBUTING.md) 
- [x] I have installed + ran pre-commit hooks.

### CODE REVIEW

File: bodo/pandas/frame.py [hunk 1]
```diff
@@ -664,7 +673,7 @@ def __setitem__(self, key, value) -> None:
             "Only setting a column with a Series created from the same dataframe is supported."
         )
 
-    @check_args_fallback(supported=["func", "axis", "args"])
+    @check_args_fallback(supported=["func", "axis"])
```
Comment (scott-routledge2) [inline bodo/pandas/frame.py:676]: Same with df.apply, changes need to be reverted

**Review by scott-routledge2 — CHANGES_REQUESTED** [review]
Looks like some of the recent changes to this file got overwritten when you merged with main.

File: bodo/pandas/frame.py [hunk 2]
```diff
@@ -552,6 +558,13 @@ def merge(
             for a, b in zip(left_on, right_on)
         ]
 
+        # TODO[BSE-4812]: support keys that are not in the beginning of the input tables
```
Comment (scott-routledge2) [inline bodo/pandas/frame.py:561]: I think this is a merge issue, revert?

File: bodo/pandas/frame.py [hunk 3]
```diff
@@ -648,11 +661,7 @@ def __setitem__(self, key, value) -> None:
                 # Update internal state
                 self._mgr._plan = new_plan
                 head_val = value._head_s
-                # Copy and update head in case reused
```
Comment (scott-routledge2) [inline bodo/pandas/frame.py:651]: Same here change needs to be reverted

File: bodo/pandas/frame.py [hunk 4]
```diff
@@ -873,10 +874,66 @@ def _get_set_column_plan(
...
+    """Check on, left_on and right_on values for type correctness
+    (currently only str, str list, str tuple, or None are supported)
+    and matching number of elements. If failed to validate, raise error.
+    Also checks membership in left and right DFs to validate keys.
+    """
+    validate_on(on)
+    validate_on(left_on)
+    validate_on(right_on)
+
+    if on is None and left_on is None and right_on is None:
```
Comment (scott-routledge2) [inline bodo/pandas/frame.py:918]: What happens if on, left_on and right_on are not specified? Do we need to check for common columns?

File: bodo/tests/test_df_lib/test_frontend.py [hunk 5]
```diff
@@ -73,3 +73,78 @@ def test_df_apply_bad_dtype_fallback_warning():
...
+
+    # Pandas merge should raise ValueError due to mismatched key lengths
+    with pytest.raises(ValueError):
+        df1.merge(df2, how="inner", left_on=["A", "B"], right_on=["C"])
+
+    # BodoDataFrame merge should raise ValueError as well
+    with pytest.raises(ValueError):
+        bdf1.merge(bdf2, how="inner", left_on=["A", "B"], right_on=["C"])
+
+    # bdf1.merge(bdf2, how="inner", left_on=["C"], right_on=["C"])
```
Comment (scott-routledge2) [inline bodo/tests/test_df_lib/test_frontend.py:107]:
```suggestion

```

**Review by scott-routledge2 — APPROVED** [review]
LGTM, thanks @chrisoh7. See my comments.

**Review by ehsantn — APPROVED** [review]
Looks good to me. Same comments as Scott. Thanks @chrisoh7 .
